url/urlsearchparams.go

```diff
@@ -0,0 +1,368 @@
+package url
+
+import (
+	"fmt"
+	"net/url"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/dop251/goja"
+)
+
+var (
+	reflectTypeString = reflect.TypeOf("")
+	reflectTypeObject = reflect.TypeOf(map[string]interface{}{})
+	reflectTypeArray  = reflect.TypeOf([]interface{}{})
+	reflectTypeMap    = reflect.TypeOf([][2]interface{}{})
+)
+
+func newInvalidTypleError(r *goja.Runtime) *goja.Object {
+	return newError(r, "ERR_MISSING_ARGS", "Each query pair must be an iterable [name, value] tuple")
+}
+
+func newMissingArgsError(r *goja.Runtime, msg string) *goja.Object {
+	return newError(r, "ERR_MISSING_ARGS", msg)
+}
+
+func newInvalidArgsError(r *goja.Runtime) *goja.Object {
+	return newError(r, "ERR_INVALID_ARG_TYPE", `The "callback" argument must be of type function. Received undefined`)
+}
+
+func newError(r *goja.Runtime, code string, msg string) *goja.Object {
+	o := r.NewTypeError("[" + code + "]: " + msg)
+	o.Set("code", r.ToValue(code))
+	return o
+}
+
+// Currently not supporting the following:
+//
+//   - ctor(iterable): Using function generators
+func createURLSearchParamsConstructor(r *goja.Runtime) goja.Value {
+	f := r.ToValue(func(call goja.ConstructorCall) *goja.Object {
+		u, _ := url.Parse("")
+		if len(call.Arguments) > 0 {
+			p := call.Arguments[0]
+			e := p.Export()
+			switch p.ExportType() {
+			case reflectTypeString:
+				u = buildParamsFromString(e.(string))
+			case reflectTypeObject:
+				u = buildParamsFromObject(e.(map[string]interface{}))
+			case reflectTypeArray:
+				u = buildParamsFromArray(r, e.([]interface{}))
+			case reflectTypeMap:
+				u = buildParamsFromMap(r, e.([][2]interface{}))
+			}
+		}
+
+		res := r.ToValue(newFromURL(u)).(*goja.Object)
+		res.SetPrototype(call.This.Prototype())
+		return res
+	}).(*goja.Object)
+
+	f.Set("prototype", createURLSearchParamsPrototype(r))
+	return f
+}
+
+// If Parsing results in a path, we move this to the RawQuery
+func buildParamsFromString(s string) *url.URL {
+	u, err := url.Parse(s)
+	if err != nil {
+		return nil
+	}
+
+	if u.Path != "" && u.RawQuery == "" {
+		v, err := url.Parse(fmt.Sprintf("?%s", u.Path))
+		if err != nil {
+			return nil
+		}
+		return v
+	}
+
+	return u
+}
+
+func buildParamsFromObject(o map[string]interface{}) *url.URL {
+	query := searchParams{}
+	for k, v := range o {
+		if val, ok := v.([]interface{}); ok {
+			vals := []string{}
+			for _, e := range val {
+				vals = append(vals, fmt.Sprintf("%v", e))
+			}
+			query = append(query, searchParam{name: k, value: vals})
+		} else {
+			query = append(query, searchParam{name: k, value: []string{fmt.Sprintf("%v", v)}})
+		}
+	}
+	u, _ := url.Parse("")
+	u.RawQuery = encodeSearchParams(query)
+	return u
+}
+
+func buildParamsFromArray(r *goja.Runtime, a []interface{}) *url.URL {
```

Comment:
Same here, consider working with *Object directly. It's a bit less convenient, but still more memory efficient.